### PR TITLE
Change HttpRequest/ResponseJsonExtensions namespace

### PR DESCRIPTION
**Pull request by halter73**
- Extension methods should be in the same namespace as the type
  they're extending
- This is new in 5.0.0-preview6 (#21731), so it's not breaking.

Addresses #23495

### CODE REVIEW

**Comment by JamesNK** [comment]
`Microsoft.AspNetCore.Http.Json` was chosen intentionally because we didn't want to automatically include our particular serialization extension methods that use S.T.J serialization to everyone.

Do we want to change that?

**Comment by halter73** [comment]
Why don't we want to automatically include the S.T.J extension methods? To avoid confusion in intellisense if someone want to use N.J extension methods instead? I don't remember that discussion.

Have we intentionally put extension methods in a namespace other than the type it's extending in other similar cases? I think the general guidance of putting extension methods in the namespace of the type they are describing is a good one.

**Comment by Tratcher** [comment]
> Have we intentionally put extension methods in a namespace other than the type it's extending in other similar cases?

Yes, occasionally. It depends on how much emphasis we are trying to put on said API.

**Comment by davidfowl** [comment]
I’m torn here. I think we should put this in the main namespace

**Comment by Tratcher** [comment]
The argument in design review was that these APIs are very opinionated ways to do JSON and we don't want to force that opinion as the one-true-way to do JSON.

**Comment by davidfowl** [comment]
I understand but I don’t think it these methods change anything wrt that. We have a built in way to do JSON and adding Json to the namespace doesn’t change that. It’s sorta artificial at this point because it’ll  be in all the docs on how to write JSON in ASP.NET Core.

**Comment by halter73** [comment]
How are these APIs very opinionated? Because people might prefer to return a JsonResult or use model binding? In that case, you're not touching the Request and Response objects anyway, so you'd never even see these APIs.

**Comment by davidfowl** [comment]
> How are these APIs very opinionated? Because people might prefer to return a JsonResult or use model binding? In that case, you're not touching the Request and Response objects anyway, so you'd never even see these APIs.

JsonResult uses the Formatter API in MVC which is replicable. Model binding is also pluggable though some of the defaults aren't replacable, the system as a whole is pluggable. Here, using Write/ReadJSON uses System.Text.Json and isn't pluggable.

**Comment by JamesNK** [comment]
Hard-coding a specific serializer is offering a strong opinion.

I don't feel strongly this. I'm ok with moving them to the main namespace if that is the consensus. If someone else wants to offer different extension methods there is nothing stopping them creating extension methods with a name like `ReadFromNewtonsoftJsonAsync` for example.

**Comment by davidfowl** [comment]
I think we should put it in the main namespace.

File: src/Http/Http.Extensions/ref/Microsoft.AspNetCore.Http.Extensions.netcoreapp.cs [hunk 1]
```diff
@@ -17,6 +17,24 @@ public static partial class HttpRequestExtensions
     {
         public static bool HasJsonContentType(this Microsoft.AspNetCore.Http.HttpRequest request) { throw null; }
```
Comment (JamesNK) [inline src/Http/Http.Extensions/ref/Microsoft.AspNetCore.Http.Extensions.netcoreapp.cs:18]: If we move these types then `HttpRequestExtensions` can be removed and `HasJsonContentType` be moved to `HttpRequestJsonExtensions`.

**Review by Tratcher — APPROVED** [review]

**Review by JamesNK — APPROVED** [review]

**Pull request merged**